src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -334,30 +340,77 @@ private void failUnpublished(final List<BatchItem> batch, final String reason) {
-            if (position == SubscriptionBase.InitialPosition.BEGIN) {
-                consumer.seekToBeginning(kafkaTPs);
-            } else if (position == SubscriptionBase.InitialPosition.END) {
-                consumer.seekToEnd(kafkaTPs);
-            } else {
-                throw new IllegalArgumentException("Bad offset specification " + position + " for topic " + topicId);
+        final SubscriptionBase.InitialPosition position = subscription.getReadFrom();
+
+        if (position == SubscriptionBase.InitialPosition.CURSORS) {
+            final List<SubscriptionCursorWithoutToken> etInitialCursors = subscription.getInitialCursors().stream()
+                    .filter(c -> c.getEventType().equals(eventType.getName()))
+                    .collect(Collectors.toList());
+            return getSubscriptionTopicInitPositions(eventType.getTopic(), etInitialCursors);
+
+        } else {
+            try (final Consumer<String, String> consumer = kafkaFactory.getConsumer()) {
+                final org.apache.kafka.common.TopicPartition[] kafkaTPs = consumer
+                        .partitionsFor(eventType.getTopic())
+                        .stream()
+                        .map(p -> new org.apache.kafka.common.TopicPartition(eventType.getTopic(), p.partition()))
+                        .toArray(org.apache.kafka.common.TopicPartition[]::new);
+                consumer.assign(Arrays.asList(kafkaTPs));
+                if (position == SubscriptionBase.InitialPosition.BEGIN) {
+                    consumer.seekToBeginning(kafkaTPs);
+                } else if (position == SubscriptionBase.InitialPosition.END) {
+                    consumer.seekToEnd(kafkaTPs);
+                } else {
+                    throw new IllegalArgumentException("Bad offset specification " + position + " for topic " +
+                            eventType.getTopic());
+                }
+                return Stream.of(kafkaTPs).collect(Collectors.toMap(
+                        tp -> String.valueOf(tp.partition()),
+                        consumer::position));
+            } catch (final Exception e) {
+                throw new ServiceUnavailableException("Error occurred when fetching partitions offsets", e);
             }
-            return Stream.of(kafkaTPs).collect(Collectors.toMap(
-                    tp -> String.valueOf(tp.partition()),
-                    consumer::position));
-        } catch (final Exception e) {
-            throw new ServiceUnavailableException("Error occurred when fetching partitions offsets", e);
         }
+    }
 
+    private Map<String, Long> getSubscriptionTopicInitPositions(final String topic,
+                                                                final List<SubscriptionCursorWithoutToken> etCursors) {
+        final Map<String, Long> positions = new HashMap<>();
+        final List<SubscriptionCursorWithoutToken> cursorsWithBegin = etCursors.stream()
```

Comment:
It's not possible to have END in a cursor, we don't have such a value.
The only place we use "end" is a "read_from" property of subscription.